### PR TITLE
Make core project "cats friendly" and get rid of Futures

### PR DESCRIPTION
**Pull request by danilbykov**
Hi Alfonso,

Thank you for your project, I was able to quickly add telegram bot in my application.

I would like to propose some changes in this pull request. Project 'core' is based on scala.concurrent.Future but it actually doesn't use any specific Future's properties. Core needs only instance of cats.MonadError. So we can abstract over Effect container, declare BotBase as
```
trait BotBase[F] {
  implicit val monad: MonadError[F, Throwable]
  ...
}
```
and use generic F instead Future.

This project is based on sttp which provides backends for Future from standard lib, IO from cats-effect, Task from Monix and IO from ZIO. This gives us compatibility with any of these libraries without necessary to convert IO to Future and Future to IO for any operation.

I also tried to get rid of mutable state inside Polling.scala and BotBase.scala:

1. Polling had variable
```
@volatile private var polling: Future[Unit]
```
which allowed to interrupt polling. But this is not needed for IO/Task which has built-in cancellation functionality. So I decided to declare common trait telegram.api.Polling and two "implementations" telegram.future.Polling for Future's and telegram.cats.api.Polling for cats-effect. New trait telegram.cats.api.Polling doesn't have mutable state.

2. BotBase had mutable variable
```
protected var getMe: User = _
```
to handle "respectful" commands. I removed this variable and pass User as argument of method BotBase.receiveUpdate. It required some changes in telegram.api.declarative.Commands.

Even though this pull request has a lot of changes and they are relatively straight-forward. I tried to minimize changes for users of this library to migrate code to this version. Basically they need just to add [Future] for traits from telegram.api.declarative, for example,
```
trait Declarative extends Updates[Future]
  with Messages[Future]
  with ChannelPosts[Future]
  with Callbacks[Future]
  with InlineQueries[Future]
  with Payments[Future]
```

### CODE REVIEW

**Comment by mukel** [comment]
I've been following you changes since the fork, looks pretty neat! I'll go for a review sprint over the weekend!

**Comment by danilbykov** [comment]
Today I tried to import all these changes in my project and realized that module 'implementation/cats' is not needed at all. This module does have actual dependencies on cats-effect (as I wanted initially) and can be moved to core module in package com.bot4s.telegram.cats. So I did this in recent commit.

Apart from that I added module with examples for Monix (actually it contains only one EchoBot for now). So now we have two new modules with examples: one for cats-effect and another for Monix.

**Comment by mukel** [comment]
I've been sidelined by my thesis deadline; please bear with me, I'll put some work to merge this once I'm done. This is major change that needs some polished docs, working examples in the README...

File: core/src/com/bot4s/telegram/future/Polling.scala [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+import com.bot4s.telegram.api.{Polling => BasePolling}
+import com.bot4s.telegram.methods.{DeleteWebhook, GetMe}
+import com.bot4s.telegram.models.{Update, User}
+import slogging.StrictLogging
+
+import scala.concurrent.Future
+import scala.util.control.NonFatal
+
+trait Polling extends BasePolling[Future] with BotExecutionContext with StrictLogging {
```
[the remark this answers is not in the document]
Comment (danilbykov) [reply core/src/com/bot4s/telegram/future/Polling.scala]: Actually I think I found better solution for this issue. I added modifier `private[telegram]` for trait `Polling`. So if someone uses this trait then he/she will get compilation error and will be forced to change to `future.Polling` or `cats.Polling`.

**Comment by danilbykov** [comment]
Sure, no problem.

File: core/src/com/bot4s/telegram/api/BotBase.scala [hunk 2]
```diff
@@ -34,36 +35,41 @@ trait BotBase {
...
-    u.shippingQuery.foreach(receiveShippingQuery)
-    u.preCheckoutQuery.foreach(receivePreCheckoutQuery)
-  }
+  protected lazy val unit = monad.pure(())
 
-  def receiveMessage(message: Message): Unit = {}
-  def receiveEditedMessage(editedMessage: Message): Unit = {}
+  def receiveMessage(message: Message): F[Unit] = unit
+  def receiveEditedMessage(editedMessage: Message): F[Unit] = unit
+  def receiveExtMessage(extMessage: (Message, Option[User])): F[Unit] = unit
```
[the remark this answers is not in the document]
Comment (mukel) [reply core/src/com/bot4s/telegram/api/BotBase.scala]: It should be the other way around, `getMe` should be called for both and the bot "identity" initialized.

**Pull request merged**